Fix crash in GCD codec in case when zeros present in data

nikitamikhaylov opened the pull request:
Closes #56672
Closes #56673

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed crash in GCD codec in case when zeros are present in data. This closes: https://github.com/ClickHouse/ClickHouse/issues/56672

nikitamikhaylov committed:
Fix crash

nikitamikhaylov committed:
Update CompressionCodecGCD.cpp

nikitamikhaylov committed:
Better

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #56704 from ClickHouse/ClickHouse:fix-gcd-crash -->
*This is an automated comment for commit bdbffda46e0cee461293a29203a7c347f931fa48 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56704/bdbffda46e0cee461293a29203a7c347f931fa48/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6855588170">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56704/bdbffda46e0cee461293a29203a7c347f931fa48/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

nikitamikhaylov committed:
Update CompressionCodecGCD.cpp

nikitamikhaylov committed:
Update CompressionCodecGCD.cpp

nikitamikhaylov committed:
Update CompressionCodecGCD.cpp

rschu1ze committed:
Some fixups (mostly for consistency)

- Made the exception texts consistent accross codecs
- Slightly less noisy comments in C++ code
- Remove unlikely() because it did not make a difference considering the
  total function costs
- Consolidated tests into existing test, delete leftover
  02869_gcd_codec_test_incorrect_type test (it was consolidated into
  02872_gcd_codec long time ago)

rschu1ze committed:
Bug: #56673

rschu1ze committed:
Try to fix stateful test

nikitamikhaylov commented:
Hmm, another crash: 
```
023.11.15 05:31:55.129180 [ 2561 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2023.11.15 05:31:55.129235 [ 2561 ] {} <Fatal> BaseDaemon: (version 23.11.1.1003, build id: D06B1D4303D3591211BA011EC5A384E94FB468FD, git hash: 49ffe85f3c70f55caa64d5079e7817e1aa1ba887) (from thread 1002) Received signal 11
2023.11.15 05:31:55.129275 [ 2561 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2023.11.15 05:31:55.129286 [ 2561 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2023.11.15 05:31:55.129299 [ 2561 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000001156d1ef 0x0000000011570651 0x000000001250e1d0 0x000000001251d5d9 0x0000000014ff8614 0x0000000014ff9811 0x0000000015104f27 0x00000000151031bc 0x00007f1504e55ac3 0x00007f1504ee6bf4
2023.11.15 05:31:55.129318 [ 2561 ] {} <Fatal> BaseDaemon: ########################################
2023.11.15 05:31:55.129327 [ 2561 ] {} <Fatal> BaseDaemon: (version 23.11.1.1003, build id: D06B1D4303D3591211BA011EC5A384E94FB468FD, git hash: 49ffe85f3c70f55caa64d5079e7817e1aa1ba887) (from thread 1002) (query_id: be9f0962-e445-4308-97bc-0f79005226b8) (query: SELECT * FROM hits_gcd FORMAT Null;) Received signal Segmentation fault (11)
2023.11.15 05:31:55.129334 [ 2561 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2023.11.15 05:31:55.129340 [ 2561 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000001156d1ef 0x0000000011570651 0x000000001250e1d0 0x000000001251d5d9 0x0000000014ff8614 0x0000000014ff9811 0x0000000015104f27 0x00000000151031bc 0x00007f1504e55ac3 0x00007f1504ee6bf4
2023.11.15 05:31:55.143574 [ 2561 ] {} <Fatal> BaseDaemon: 2.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:174: std::__shared_count::__release_shared[abi:v15000]()
2023.11.15 05:31:55.143598 [ 2561 ] {} <Fatal> BaseDaemon: 2.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:215: std::__shared_weak_count::__release_shared[abi:v15000]()
2023.11.15 05:31:55.143605 [ 2561 ] {} <Fatal> BaseDaemon: 2.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:702: ~shared_ptr
2023.11.15 05:31:55.143612 [ 2561 ] {} <Fatal> BaseDaemon: 2.4. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:768: std::shared_ptr<DB::ThreadGroup>::reset[abi:v15000]()
2023.11.15 05:31:55.143622 [ 2561 ] {} <Fatal> BaseDaemon: 2. ./build_docker/./src/Interpreters/ThreadStatusExt.cpp:273: DB::ThreadStatus::detachFromGroup() @ 0x000000001156d1ef in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.154851 [ 2561 ] {} <Fatal> BaseDaemon: 3. ./build_docker/./src/Interpreters/ThreadStatusExt.cpp:676: DB::CurrentThread::QueryScope::~QueryScope() @ 0x0000000011570651 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.187122 [ 2561 ] {} <Fatal> BaseDaemon: 4.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/optional:288: std::__optional_destruct_base<DB::CurrentThread::QueryScope, false>::reset[abi:v15000]()
2023.11.15 05:31:55.187150 [ 2561 ] {} <Fatal> BaseDaemon: 4. ./build_docker/./src/Server/TCPHandler.cpp:586: DB::TCPHandler::runImpl() @ 0x000000001250e1d0 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.237313 [ 2561 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Server/TCPHandler.cpp:2257: DB::TCPHandler::run() @ 0x000000001251d5d9 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.238621 [ 2561 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x0000000014ff8614 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.240601 [ 2561 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<Poco::Net::TCPServerConnection>::operator()[abi:v15000](Poco::Net::TCPServerConnection*) const
2023.11.15 05:31:55.240617 [ 2561 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<Poco::Net::TCPServerConnection, std::default_delete<Poco::Net::TCPServerConnection>>::reset[abi:v15000](Poco::Net::TCPServerConnection*)
2023.11.15 05:31:55.240624 [ 2561 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2023.11.15 05:31:55.240628 [ 2561 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x0000000014ff9811 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.242823 [ 2561 ] {} <Fatal> BaseDaemon: 8. ./build_docker/./base/poco/Foundation/src/ThreadPool.cpp:202: Poco::PooledThread::run() @ 0x0000000015104f27 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.244693 [ 2561 ] {} <Fatal> BaseDaemon: 9.1. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:231: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::get()
2023.11.15 05:31:55.244705 [ 2561 ] {} <Fatal> BaseDaemon: 9.2. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::assign(Poco::Runnable*)
2023.11.15 05:31:55.244711 [ 2561 ] {} <Fatal> BaseDaemon: 9.3. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:180: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::operator=(Poco::Runnable*)
2023.11.15 05:31:55.244716 [ 2561 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:350: Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000151031bc in /usr/lib/debug/usr/bin/clickhouse.debug
2023.11.15 05:31:55.244725 [ 2561 ] {} <Fatal> BaseDaemon: 10. ? @ 0x00007f1504e55ac3 in ?
2023.11.15 05:31:55.244729 [ 2561 ] {} <Fatal> BaseDaemon: 11. ? @ 0x00007f1504ee6bf4 in ?
2023.11.15 05:31:55.244735 [ 2561 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2023.11.15 05:31:55.350413 [ 2561 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2023.11.15 05:31:55.350533 [ 2561 ] {} <Fatal> BaseDaemon: Changed settings: min_compress_block_size = 3044968, max_compress_block_size = 669313, max_block_size = 99022, max_threads = 49, max_read_buffer_size = 533906, connect_timeout_with_failover_ms = 2000, connect_timeout_with_failover_secure_ms = 3000, idle_connection_timeout = 36000, s3_check_objects_after_upload = true, use_uncompressed_cache = true, stream_like_engine_allow_direct_select = true, enable_multiple_prewhere_read_steps = false, replication_wait_for_inactive_replica_timeout = 30, min_count_to_compile_sort_description = 0, group_by_two_level_threshold = 928801, group_by_two_level_threshold_bytes = 15712951, distributed_aggregation_memory_efficient = false, enable_memory_bound_merging_of_aggregation_results = false, allow_nonconst_timezone_arguments = true, input_format_parallel_parsing = false, min_chunk_bytes_for_parallel_parsing = 8781555, min_bytes_to_use_direct_io = 8156827361, min_bytes_to_use_mmap_io = 10737418240, log_queries = true, insert_quorum_timeout = 60000, http_wait_end_of_query = true, http_response_buffer_size = 5531955, fsync_metadata = false, http_send_timeout = 60., http_receive_timeout = 60., opentelemetry_start_trace_probability = 0.10000000149011612, max_untracked_memory = 1048576, memory_profiler_step = 1048576, log_comment = '00178_gcd_codec.sql', send_logs_level = 'warning', prefer_localhost_replica = false, optimize_read_in_order = false, aggregation_in_order_max_block_bytes = 34420084, read_in_order_two_level_merge_threshold = 51, allow_introspection_functions = true, database_atomic_wait_for_drop_and_detach_synchronously = true, local_filesystem_read_method = 'mmap', remote_filesystem_read_method = 'read', local_filesystem_read_prefetch = true, remote_filesystem_read_prefetch = false, async_insert_busy_timeout_ms = 5000, enable_filesystem_cache = true, enable_filesystem_cache_on_write_operations = true, load_marks_asynchronously = true, allow_prefetched_read_pool_for_remote_filesystem = false, allow_prefetched_read_pool_for_local_filesystem = false, filesystem_prefetch_min_bytes_for_single_read_task = 1048576, filesystem_prefetch_max_memory_usage = 67108864, filesystem_prefetches_limit = 10, optimize_sorting_by_input_stream_properties = false, insert_keeper_max_retries = 20, insert_keeper_fault_injection_probability = 0.009999999776482582, session_timezone = 'America/Mazatlan'
2023.11.15 05:31:59.392726 [ 998 ] {} <Fatal> Application: Child process was terminated by signal 11.

```

rschu1ze commented:
Unrelated for sure. The only (very remotely) related crash I could find was https://github.com/ClickHouse/ClickHouse/pull/45236#issuecomment-1508392520

rschu1ze committed:
Undo duplicate test that also times out

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into fix-gcd-crash
